Lukaszgryglicki 2368 do merge if release note needed

lukaszgryglicki opened the pull request:
Fixes #2368.

Copied from here: https://github.com/kubernetes/test-infra/pull/2456#issuecomment-293693766)
- make sure that munger will not merge if there is `release-note-label-needed` Label
- release-note munger will not apply `do-not-merge`

Some details:
- Change to `mungegithub/github/github.go`:`WriteComment`, when testing and creating comments we're not adding them to `obj.comments` like all other functions do. So then when testing for added comment via `obj.ListComments()` we cannot get them back. This happens in testing environment, in non-test environments real github calls are used. This is exactly the same as with `AddLabel`, `RemoveLabel`, `HasLabel` etc. - all of them are failing in test environment (on real GH call) and puts that info in logs, but they update local `obj` variables so we still can do testing.
- `mungegithub/mungers/fsm/prereview.go` I've found this typo while working, so I've corrected it.
- `mungegithub/mungers/release-note-label.go` no longer adds `do-not-merge` label. It only adds slightly modified comment (coverage is extended to test for this comment). So its logic changed a little. It was adding `do-not-merge`label & comment only when LGTM label was present. Now it adds comment when LGTM is present and only if we just added `release-note-label-needed` in this call (to avoid adding it multiple times)
- `mungegithub/mungers/submit-queue.go` now checks if there is a label `release-note-label-needed`. If so it is not merging and puts comment about this. Coverage also updated.

cblecker commented:
@k8s-bot ok to test
/assign

Sane enough to test. Will review for feedback.

cblecker commented on mungegithub/mungers/fsm/prereview.go:
```diff
@@ -28,7 +28,7 @@ const (
 	releaseNoteActionRequired = "release-note-action-required"
 	releaseNoteExperimental   = "release-note-experimental"
-	releaseNoteLabelNeeded    = "release-note-none"
+	releaseNoteNone           = "release-note-none"
```
both releaseNoteLabelNeeded and releaseNoteNone should be here, shouldn't they?

lukaszgryglicki replied:
I think not.
We are checking there is there is one of 4 required releaseNote labels.
And releaseNoteLabelNeeded is of other kind. It is to mark that we're missing one of those 4....

cblecker replied:
Ah okay. Yeah, these are just the constant definitions, but `releaseNoteLabelNeeded` isn't called in this particular piece of code.

cblecker commented on mungegithub/mungers/submit-queue_test.go:
```diff
@@ -523,6 +527,22 @@ func TestSubmitQueue(t *testing.T) {
 			isMerged:        true,
 		},
 		{
+			name:            "Test1+releaseNoteNeededLebel",
```
nit: typo

lukaszgryglicki replied:
Lebel - got it. Will fix

cblecker commented on mungegithub/mungers/submit-queue.go:
```diff
@@ -896,6 +896,7 @@ const (
 	unmergeable             = "PR is unable to be automatically merged. Needs rebase."
 	undeterminedMergability = "Unable to determine is PR is mergeable. Will try again later."
 	noMerge                 = "Will not auto merge because " + doNotMergeLabel + " is present"
+	noReleaseNote           = "Will not auto merge because " + releaseNoteLabelNeeded + " is present. You need to set one of the release-note* labels."
```
I think we should leave off the "You need to set.." bit as we'll have a comment on the PR, and a string that long won't show up properly in the Github UI.

lukaszgryglicki replied:
Will do.

cblecker commented on mungegithub/mungers/release-note-label_test.go:
```diff
@@ -38,107 +39,120 @@ var (
-			issue:       github_test.Issue(botName, 1, []string{lgtmLabel}, true),
-			mustHave:    []string{lgtmLabel, doNotMergeLabel, releaseNoteLabelNeeded},
-			mustNotHave: []string{},
+			name:             "LGTM only",
```
Why is this test changing?

lukaszgryglicki replied:
Due to `gofmt` after I've added property check `mustHaveComment1: true` - this forced to few lines above it to have more whitespace.

cblecker replied:
No I mean that `doNotMergeLabel` is moving from `mustHave` to `mustNotHave`

lukaszgryglicki replied:
This is what that PR does. `release-note` munger now should never add `do-not-merge` label.

cblecker replied:
but we're not explicitly setting or removing it. I don't think it should be in the `mustNotHave` as a result, because there are other conditions (maybe not the munger) that may set it. `mustNotHave` is more for, we're explicitly removing it, so we need to verify it worked.

cblecker commented on mungegithub/mungers/release-note-label_test.go:
```diff
@@ -38,107 +39,120 @@ var (
-			name:     "No labels",
-			issue:    github_test.Issue(botName, 1, []string{}, true),
-			mustHave: []string{releaseNoteLabelNeeded},
+			name:        "No labels",
```
Why is this test changing?

lukaszgryglicki replied:
Same as above:
```
Due to `gofmt` after I've added property check `mustHaveComment1: true` - this forced to few lines above it to have more whitespace.
```

cblecker commented on mungegithub/mungers/release-note-label_test.go:
```diff
@@ -38,107 +39,120 @@ var (
 			issue:       github_test.Issue(botName, 1, []string{lgtmLabel, releaseNote}, true),
 			mustHave:    []string{lgtmLabel, releaseNote},
-			mustNotHave: []string{releaseNoteLabelNeeded},
+			mustNotHave: []string{releaseNoteLabelNeeded, doNotMergeLabel},
```
Do we really need to test that it doesn't have `do-not-merge`? Issues could have `do-not-merge` for other reasons than release note issues.

lukaszgryglicki replied:
Well I've added this to ensure that munger is NOT adding this label.
It is now supposed to never add it, so I've updated a bunch of tests.
Anyway - I need to check for that in at least one test.

grodrigues3 commented:
As with the other PR, please squash commits when you're ready for merge.

lukaszgryglicki commented:
Not sure why it failed. I've squashed commits, compared manually to see if nothing is broken. Tests pass locally..... is it some flaky test / intermittent failure?
All tests pass:
```
bazel test //mungegithub/...
bazel test //prow/...
```

Maybe re-run? Or maybe due to merge with upstream something fails in other tests?

lukaszgryglicki commented:
@k8s-bot bazel test this

lukaszgryglicki commented:
No idea what's wrong, I think it is unrelated to this PR:
```
 4) Clusters refilter shows PRs when demanded:
     TypeError: Cannot read property 'length' of undefined
      at sum (/workspace/k8s.io/test-infra/triage/model.js:69:10)
      at clustersSum (/workspace/k8s.io/test-infra/triage/model.js:75:10)
      at Clusters.sum.c (/workspace/k8s.io/test-infra/triage/model.js:128:36)
      at Array.map (native)
      at sum (/workspace/k8s.io/test-infra/triage/model.js:71:14)
      at new Clusters (/workspace/k8s.io/test-infra/triage/model.js:128:16)
      at Context.<anonymous> (/workspace/k8s.io/test-infra/triage/script_test.js:20:25)
```

spxtr commented:
You're right, looks like #2491 broke it.

lukaszgryglicki commented:
Thx, need to switch for other tasks, so I'll be back after Holidays.

fejta commented:
@k8s-bot bazel test this

lukaszgryglicki commented:
I've removed doNotMerge test from most places. I've only left that test where I think it should be tested (like when there is `release-not-label-needed` to ensure that munger won't set `do-not-merge` label.
So there are still few places where I've moved `do-not-merge` label from expected to unexpected - this is exactly what this PR does.

cblecker commented on mungegithub/mungers/release-note-label.go:
```diff
@@ -144,20 +144,15 @@ func (r *ReleaseNoteLabel) Munge(obj *github.MungeObject) {
-	}
-
-	obj.WriteComment(releaseNoteBody)
-	obj.AddLabel(doNotMergeLabel)
```
Looking at this logic, I don't think any of these changes are needed, other than removing this line.

lukaszgryglicki replied:
If I only remove `obj.AddLabel(doNotMergeLabel)` then if just 4 lines above:
```
if obj.HasLabel(doNotMergeLabel) {
   return
}
```

Will not cause return, so we will be adding `obj.WriteComment(releaseNoteBody)` comment every time.
This is why I've added code to ensure that we're adding this comment only when we also added `obj.AddLabel(releaseNoteLabelNeeded)` at the same time.

cblecker replied:
return is an escape out of this function. if the PR has do-not-merge for some other reason, then it's not eligible for merge and shouldn't write the comment

lukaszgryglicki replied:
And what if PR doesn't have `do-not-merge` and we are no longer setting it there.
In that case in 1st call of `Muge` we will add `release-note-label-needed` and comment on PR (no more adding `do-not-merge`, but then:
2nd call (please follow Munge function).
It will check if object have one of 4 release note labels. It doesn't.
It will check if `release-note-label-needed` is present - it is.
Then it will hit `if obj.HasLabel(doNotMergeLabel)` but now we didn't added that label in 1st call so it will proceed and add comment agaig.
The same in 3rd call, 4th call etc.
Each `Munge` call will add `obj.WriteComment(releaseNoteBody`

Before removing single line `obj.AddLabel(doNotMergeLabel)` it was adding `do-not-merge` label in 1st call, so any next call will check `if obj.HasLabel(doNotMergeLabel) {` and then return - so it won't add comment more than once.

So I think my changes are actually OK.
Because I'm adding additional condition to put comment only in the same call when `release-note-lable-needed` were added.

If I'm wrong please let me know - but please describe why I'm wrong, this is my 2nd PR, and maybe I missed something...
But please see this file:
https://github.com/kubernetes/test-infra/blob/master/mungegithub/mungers/release-note-label.go
And `Munge` function.
Please remove only one line (as You wanted me to do) remove: `obj.AddLabel(doNotMergeLabel)` (last line)
And then imagine what will happen if we call `Munge` 5 times. Comment `obj.WriteComment(releaseNoteBody)` would be added 5 times then.
Then please take a look on my implementation in this PR.
My implementation will not add `do-not-merge` label but also wont' add `obj.WriteComment(releaseNoteBody)` more than once.

If I'm wrong please correct me.... what exactly I'm assuming bad.
Maybe `Munge` function is called only in specific conditions? Myabe this is what I missed?
BTW: when `Munge` is called? (for this particular munger)?
I know that `prow` reacts to Github web hooks.... how about mungegithub? Periodic?

cblecker replied:
Okay, I see what you mean. The trip wire to ensure that this function didn't call multiple times was if this function ran all the way to the end, the PR had `do-not-merge` added to it.

The issue I'm seeing with the current logic is the `releaseNoteLabelNeededAdded` variable won't carry over multiple runs of this function. So we'll still be in a place that if the function is called, it will add the note over and over.

lukaszgryglicki replied:
Yes but this variable is used only when there was no `release-not-label-needed` yet and then we're adding label `release-note-label-needed` (1st call) - then we want to add comment.
Next call (2nd and others) won't add comment because there will be `release-note-label-needed` already comment won't be added again, see:
```
	releaseNoteLabelNeededAdded := false
	if !obj.HasLabel(releaseNoteLabelNeeded) {
		obj.AddLabel(releaseNoteLabelNeeded)
		releaseNoteLabelNeededAdded = true
	}

	if releaseNoteLabelNeededAdded && obj.HasLabel(lgtmLabel) {
		obj.WriteComment(releaseNoteBody)
	}
```

cblecker replied:
Then the comment will almost never be added. The first run will set the variable to false, then see there isn't the needed label, and then add it. But the PR won't have LGTM so the comment won't be written. Then the next run will have the variable set to false, the PR *does* have the needed label so the variable won't be set to true, and then even if the PR has LGTM, the variable won't be true. Still no comment.

lukaszgryglicki replied:
This is true.... so maybe we should talk when exactly this comment should be added?
My ide was to add it in the same time when release-note-label-needed.
Maybe just remove condition for LGTM label?
I'm open to ideas.... I just needed to change some logic here (not relay on do-not-merge label), an dthat was my idea (may be wrong).

cblecker replied:
Two possible solutions:
- Use `!obj.HasLabel(releaseNoteLabelNeeded)` as the tracking mechanism and as soon as we poll for release note, if the PR doesn't have the right label, add the comment in the PR right away.
- Only add the note at LGTM time, but remove the LGTM if we don't have the right releasenote process followed. If we did this, we could simplify things further by not needing to make release-note-label-needed a blocking label, because removing the LGTM would have the same effect.

lukaszgryglicki replied:
OK moved to my TODO list, because I have to switch to other tasks for next few days.

cblecker commented:
@kubernetes/sig-contributor-experience-misc FYI on this. slight change to how release note process works with a couple options. details here: https://github.com/kubernetes/test-infra/pull/2478#discussion_r113542067

cblecker commented:
@lukaszgryglicki Hey there! Any update on this? :)

cblecker commented:
@lukaszgryglicki ping

cblecker commented:
@lukaszgryglicki are you still planning on working on this PR?

lukaszgryglicki commented:
I was not able to work on this, I was directed to totally different stuff. Sorry.
I didn't notice those posts in a ton of kubernetes email/day :/

cblecker commented:
Fixed by #4196
/close

cblecker commented:
p.s. thanks @stevekuznetsov !
